pkg/sigstorebundle/store.go

```diff
@@ -0,0 +1,279 @@
+	var bundleLimits *BundleLimits
+	if len(limits) > 0 && limits[0] != nil {
+		bundleLimits = limits[0]
+	} else {
+		bundleLimits = DefaultBundleLimits()
+	}
+	if bundle == nil {
+		return nil, fmt.Errorf("bundle is nil")
+	}
+
+	if bundle.DsseEnvelope == nil {
+		return nil, fmt.Errorf("bundle missing required field: dsseEnvelope")
+	}
+
+	if bundle.DsseEnvelope.Payload == "" {
+		return nil, fmt.Errorf("dsseEnvelope.payload is empty")
+	}
+
+	// Check payload size before decoding (base64 encoded size * 3/4 ≈ decoded size)
+	estimatedSize := len(bundle.DsseEnvelope.Payload) * 3 / 4
+	maxPayloadSizeBytes := bundleLimits.MaxPayloadSizeMB * 1024 * 1024
+	if estimatedSize > maxPayloadSizeBytes {
+		return nil, fmt.Errorf("payload size (%d bytes) exceeds maximum allowed size (%d MB)", estimatedSize, bundleLimits.MaxPayloadSizeMB)
+	}
+
+	// Decode payload
+	payload, err := base64.StdEncoding.DecodeString(bundle.DsseEnvelope.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode dsseEnvelope.payload: %w", err)
+	}
+
+	if len(payload) == 0 {
+		return nil, fmt.Errorf("decoded payload is empty")
+	}
+
+	env := &dsse.Envelope{
+		Payload:     payload,
+		PayloadType: bundle.DsseEnvelope.PayloadType,
+	}
+
+	if len(bundle.DsseEnvelope.Signatures) == 0 {
+		return nil, fmt.Errorf("bundle has no signatures")
+	}
+
+	// Check signature count limit to prevent resource exhaustion
+	if len(bundle.DsseEnvelope.Signatures) > bundleLimits.MaxSignaturesPerBundle {
+		return nil, fmt.Errorf("bundle has %d signatures, exceeds maximum allowed (%d)", len(bundle.DsseEnvelope.Signatures), bundleLimits.MaxSignaturesPerBundle)
+	}
+
+	// Map signatures with VerificationMaterial
+	for idx, bundleSig := range bundle.DsseEnvelope.Signatures {
+		if bundleSig.Sig == "" {
+			return nil, fmt.Errorf("signature %d missing sig field", idx)
+		}
+
+		sig, err := base64.StdEncoding.DecodeString(bundleSig.Sig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode signature %d: %w", idx, err)
+		}
+
+		if len(sig) == 0 {
+			return nil, fmt.Errorf("decoded signature %d is empty", idx)
+		}
+
+		witnessSig := dsse.Signature{
+			KeyID:     bundleSig.KeyID,
+			Signature: sig,
+		}
+
+		// Map VerificationMaterial to signature fields
+		if bundle.VerificationMaterial != nil {
+			mapVerificationMaterial(&witnessSig, bundle.VerificationMaterial)
+		}
+
+		env.Signatures = append(env.Signatures, witnessSig)
+	}
+
+	return env, nil
+}
+
+// mapVerificationMaterial maps Sigstore VerificationMaterial to go-witness signature fields
+func mapVerificationMaterial(sig *dsse.Signature, vm *VerificationMaterial) {
+	if sig == nil || vm == nil {
+		return
+	}
+
+	// Certificate chain
+	if vm.X509CertificateChain != nil && len(vm.X509CertificateChain.Certificates) > 0 {
+		// First cert is leaf
+		if vm.X509CertificateChain.Certificates[0].RawBytes != "" {
+			if cert, err := base64.StdEncoding.DecodeString(vm.X509CertificateChain.Certificates[0].RawBytes); err == nil {
```

Comment:
looks like we are silently dropping when base64 decoding fails. An attacker could corrupt the data mid flight, causing data to be stored without the certificate chains. We should fail the upload when cert chains or timestamps are corrupted.